### ISSUE TITLE
Cannot reproduce the result

### ISSUE THREAD

**Issue by TonyXuQAQ** [songtaohe/LaneExtraction#9]
Dear Songtao,

Hi, I recently tried the released code of "Lane-Level Street Map Extraction from Aerial Imagery" on github, nice work! And thank you for releasing the dataset and the code.

However, the inference codes of turningLaneExtraction and turningLaneValidation are missing, so I re-implemented them by myself. I cannot reproduce the results provided in your paper, and here are some problems that I encountered, could you please give me some advice or hints? Thank you so much for your help.

1. For endpoint extraction, I followed instructions in your paper by (1) obtain the skeleton of the segmentation of straight lane centerlines, and (2) extract endpoints of these skeletons. Here is one sample direction map I obtained:      
![Screenshot from 2022-08-04 10-37-25](https://user-images.githubusercontent.com/33851435/182753326-96f8bdbb-26c5-47bf-963c-d0d45fb9e942.png)

Red points are obtained endpoints. As you can see, there are incorrect disconnections on the straight-lane centerlines. But the general performance is fine. Do my results look good to you? 

2. Based on obtained endpoints, I run inference of turnLaneValidation. However, here comes the problem: since you train the LinkModel by ground-truth direction maps, which are quite different from the predicted ones, the model presents very bad performance. It ignores almost all turning lanes. For example:
 ![Screenshot from 2022-08-04 10-47-57](https://user-images.githubusercontent.com/33851435/182753299-cebe6b61-3324-4527-8e67-f4ef99c4a84c.png)
Cyan points are node A and node B defined in the paper. These two points should be connected by a turning lane. But the model outputs a very strong reject to this connection ( the softmax result is: [1.4065192033285712e-07, 0.9999998807907104], indicating that the model is very sure that these two points should not be connected). From my perspective, the gap between the training data and inference data leads to this problem. How do you solve this? 

Thank you so much for your kind help, and looking forward to your reply!

Best regards

**Comment by TonyXuQAQ** [comment]
Here is the reply email of the author:

Hi,

Thanks for your interest in this work! 

For 1. I think the results you showed look reasonable to me. The segmentation model may produce noisy outputs, and the small dataset may worsen it. 

For 2. The quality of the segmentation output from the previous stage will affect the performance here. Another thing I found important in my experiment is the performance of the reachable lane extraction module. To debug, maybe you can check if the output from the reachable lane extraction module makes sense. It should produce good segmentation for the reachable lanes from the two terminal points. 

Please let me know if there are any further issues. Thanks!


Best,
Songtao

**Comment by TonyXuQAQ** [comment]
Dear Songtao,

Thank you so much for your information! Based on your advice, I double-checked the validation module and the extraction module. However, the inference performance is still far from expected.

For my re-produced results, here are the visualizations:

input connector 1
![Screenshot from 2022-08-09 12-08-59](https://user-images.githubusercontent.com/33851435/183563711-d37c77df-7bfd-4cca-b9c8-c0e16c5d9f84.png)

input connector 2
![Screenshot from 2022-08-09 12-09-06](https://user-images.githubusercontent.com/33851435/183563732-116dcd97-de4d-44f9-af96-d750bb6e533f.png)

input predicted direction map (after skeletonization. Cyan points are only for visualization, and they do not appear in the input of the model)
![Screenshot from 2022-08-09 12-09-19](https://user-images.githubusercontent.com/33851435/183563741-0b722f15-1efc-4c78-92b5-f76e8294c98e.png)


output reachable lane of connector 1
![Screenshot from 2022-08-09 12-09-39](https://user-images.githubusercontent.com/33851435/183563751-86b5660c-d6cd-49b1-bdfd-be73959cc0e7.png)

 
output reachable lane of connector 2
![Screenshot from 2022-08-09 12-09-47](https://user-images.githubusercontent.com/33851435/183563756-506883c1-0dc9-4888-a8dd-c387b411e0cd.png)

output extraction turning lanes
![Screenshot from 2022-08-09 12-09-58](https://user-images.githubusercontent.com/33851435/183563838-14707f2a-4c60-45dd-abce-ad78e3c04079.png)


I trained all models based on your open-sourced code and data for 500 epochs, and the validation results look fine. I think the problems are :

1. Noisy straight lane segmentation.

2. My customized way of skeletonization and endpoint extraction is not good.

So I wonder, if possible, whether you could provide the source inference code of skeletonization, turningLaneValidation, and turningLaneExtraction? It OK if the code is not cleaned and organized for open-source, I just hope to understand your approach with more details. Really thank you for your time and sincere help! 

Thanks again, and looking forward to your reply!

Best regards,

**Comment by songtaohe** [comment]
Hi @TonyXuQAQ , sorry about the bad robustness of the code. One thing in my mind is to train only the reachable lane extraction part first. That should give you a good reachable lane prediction if there is no bug in the public code. After this, you can try loading the weights of the reachable lane extraction part and training the joint model from there end-to-end. 

In the meantime, I uploaded the inference and evaluation code to the _code_for_reference_untested_ folder. I just copied and pasted the original code there, so they are not compatible with the rest of the repo.

**Comment by TonyXuQAQ** [comment]
Thank you so much for your quick reply! I will check your source code to debug. Really appreciate your effort and time.

**Comment by TonyXuQAQ** [comment]
Based on the inference scripts provided by the author, I successfully obtained the expected outputs, even though I cannot run evaluation code due to the missing of some labels. Thanks the Songtao again for your kind help and information!

 Here are some results:
Input connector A:
![Screenshot from 2022-08-11 10-03-03](https://user-images.githubusercontent.com/33851435/184052859-02bd427d-c47d-45fd-9b18-05a31ba6b989.png)
Input connector B:
![Screenshot from 2022-08-11 10-03-10](https://user-images.githubusercontent.com/33851435/184052876-791ed5f8-6980-4039-a179-dc2c1800ba2d.png)
Input direction prediction:
![Screenshot from 2022-08-11 10-03-21](https://user-images.githubusercontent.com/33851435/184052890-bec48798-d0f0-4927-b8da-ca6148cd9a15.png)
Output reachable lanes of node A
![Screenshot from 2022-08-11 10-03-27](https://user-images.githubusercontent.com/33851435/184052920-152bbf9b-9749-4625-8ae2-fb5201d8568b.png)
Output reachable lanes of node B
![Screenshot from 2022-08-11 10-03-33](https://user-images.githubusercontent.com/33851435/184052947-270c0caa-821d-4682-9a38-23bfa23a54f5.png)
Output connection
![Screenshot from 2022-08-11 10-03-39](https://user-images.githubusercontent.com/33851435/184052963-ed50e55c-d4c4-4cef-a09c-defcdde389f5.png)

Thanks the author again for your help!
